src/tests/ftest/util/slurm_utils.py

```diff
@@ -0,0 +1,194 @@
+        For very simple situations where you just want to run a simple
+        shell command via slurm this will do the trick.
+
+        cmd  --command that can be run from a shell
+        name --job name
+
+        returns --the job ID, which is used as a handle for other functions
+    """
+
+    daos_test_job = {"wrap": cmd, "job_name": name}
+    jobid = pyslurm.job().submit_batch_job(daos_test_job)
+    return jobid
+
+def run_slurm_script(script):
+
+    """
+        For running scripts, for example those created with write_slurm_script
+        function above.
+
+        script --a script file suitable to be run by slurm
+
+        returns --the job ID, which is used as a handle for other functions
+    """
+
+    cmd = "sbatch " + script
+    output = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True)
+    print output
+    if 'Submitted batch job' in output:
+        output_list = output.split()
+        if len(output_list) >= 4:
+            return output_list[3]
+
+    raise SlurmFailed("Batch job failed to start: {}".format(output))
+
+def get_job_output(handle, path):
+    """
+        Get the output from a slurm job.
+
+        handle   --slurm job id
+
+
+        returns  --string containing both stdout and stderr
+    """
+    pass
+
+def check_slurm_job(handle):
+    """
+        Get the state of a job initiated via slurm.
+
+        handle   --slurm job id
+
+        returns  --one of the slurm defined JOB_STATE_CODES strings plus
+                   one extra UNKNOWN if the handle doesn't match a known
+                   slurm job.
+    """
+
+    job_attributes = pyslurm.job().find_id(handle)
+    if job_attributes and len(job_attributes) > 0:
+        state = job_attributes[0]["job_state"]
+    else:
+        state = "UNKNOWN"
+    return state
+
+
+def register_for_job_results(handle, cb_func, maxwait=3600):
+    """
+        Register a callback for a slurm soak test job.
+
+        handle   --slurm job id
+        maxwait  --maximum time to wait in seconds, defaults to 1 hour
+        returns  --None
+
+    """
+    params = {"handle":handle, "maxwait":maxwait, "cb_func":cb_func}
+    athread = threading.Thread(target=watch_job,
+                         kwargs=params)
+    athread.start()
+
+def watch_job(handle, maxwait, cb_func):
+    """
+    This function waits for a slurm job to finish and
+    calls the provided callback function with the result.
+
+    handle   --the slurm job handle
+    maxwait  --max time in seconds to wait
+    cb_func  --whom to notify when its done
+
+    return   --none, all results handled through callback
+    """
+
+    wait_time = 0
+    while True:
+        state = check_slurm_job(handle)
+        if state == "PENDING" or state == "RUNNING":
+            if wait_time > maxwait:
+                state = "MAXWAITREACHED"
+                print("reached maxwait")
+                break
+            else:
+                wait_time += 5
+                time.sleep(5)
```

Comment:
Not sure that it matters, but just be aware that this function can actually wait longer than `maxwait` since the execution time of the function (i.e. the time spent processing the loop) is added to all of the sleeps.

If you want to actually only wait `maxwait` you can take a timestamp a the beginning of the function and then compare that to the timestamp at each iteration, rather than using a time accumulator like `wait_time`.  You would then use that timestamp to calculate your sleep until the next interval (i.e. every 5 seconds since the timestamp was taken) which will usually be 5 seconds.  But it might be that if enough execution time accumulates you only sleep 4 seconds every now and then.

Reply:
that tiny bit of time seems okay to me